distributed/scheduler.py

```diff
@@ -5861,29 +5919,43 @@ async def _rebalance_move_data(
         self, msgs: "list[tuple[WorkerState, WorkerState, TaskState]]"
     ) -> dict:
         """Perform the actual transfer of data across the network in rebalance().
-        Takes in input the output of _rebalance_find_msgs().
+        Takes in input the output of _rebalance_find_msgs(), that is a list of tuples:
+
+        - sender worker
+        - recipient worker
+        - task to be transferred
 
         FIXME this method is not robust when the cluster is not idle.
         """
-        ts: TaskState
         snd_ws: WorkerState
         rec_ws: WorkerState
+        ts: TaskState
 
         to_recipients = defaultdict(lambda: defaultdict(list))
-        to_senders = defaultdict(list)
-        for sender, recipient, ts in msgs:
-            to_recipients[recipient.address][ts._key].append(sender.address)
-            to_senders[sender.address].append(ts._key)
-
-        result = await asyncio.gather(
-            *(
-                retry_operation(self.rpc(addr=r).gather, who_has=v)
-                for r, v in to_recipients.items()
+        for snd_ws, rec_ws, ts in msgs:
+            to_recipients[rec_ws.address][ts._key].append(snd_ws.address)
+        failed_keys_by_recipient = dict(
+            zip(
+                to_recipients,
+                await asyncio.gather(
+                    *(
+                        self._gather_on_worker(w, who_has)
+                        for w, who_has in to_recipients.items()
+                    )
+                ),
             )
         )
+
+        to_senders = defaultdict(list)
+        for snd_ws, rec_ws, ts in msgs:
+            if ts._key not in failed_keys_by_recipient[rec_ws.address]:
+                to_senders[snd_ws.address].append(ts._key)
+        await asyncio.gather(
+            *(self._delete_worker_data(r, v) for r, v in to_senders.items())
+        )
```

Comment:
what happens to dependents of the to-be-deleted keys on the worker? IIUC the worker state machine is currently not equipped to deal with the necessary transitions required for something like this. The necessary transitions on worker side would be 

```
transition(ts, "memory" -> "fetch")
for dep in ts.dependents: transition(dep, "ready" -> "waiting")
```

The delete/free/release on worker side is not that sophisticated. I remember `KeyError`s popping up while debugging the deadlocks recently if I removed keys too eagerly

Reply:
Resilience of computation to having needed keys removed from under its feet is out of scope for this PR. As of this PR, it is *still* not robust to run rebalance() during a compute, in the sense that rebalance will no longer crash but compute still will.